lib/circuitry/concerns/async.rb

```diff
@@ -1,17 +1,67 @@
+require 'circuitry/processors/batcher'
+require 'circuitry/processors/forker'
+require 'circuitry/processors/threader'
+
 module Circuitry
   class NotSupportedError < StandardError; end
 
   module Concerns
     module Async
+      attr_reader :async
+
+      def self.included(base)
+        base.extend(ClassMethods)
+      end
+
+      module ClassMethods
+        def async_strategies
+          [:fork, :thread, :batch]
+        end
+
+        def default_async_strategy
+          raise NotImplementedError, "#{name} must implement class method `default_async_strategy`"
+        end
+      end
+
       def process_asynchronously(&block)
-        raise NotSupportedError, 'Your platform does not support forking' unless platform_supports_async?
+        send(:"process_via_#{async}", &block)
+      end
+
+      def async=(value)
+        value = case value
+          when false, nil then false
+          when true then self.class.default_async_strategy
+          when *self.class.async_strategies then value
+          else raise ArgumentError, "Invalid value `#{value.inspect}`, must be one of #{[true, false].concat(self.class.async_strategies).inspect}"
+        end
+
+        if value == :fork && !platform_supports_forking?
+          raise NotSupportedError, 'Your platform does not support forking'
+        end
+
+        @async = value
+      end
+
+      def async?
+        !!async
+      end
+
+      private
+
+      def platform_supports_forking?
+        Process.respond_to?(:fork)
+      end
+
+      def process_via_fork(&block)
+        Processors::Forker.process(&block)
+      end
 
-        pid = fork(&block)
-        Process.detach(pid)
+      def process_via_thread(&block)
+        Processors::Threader.process(&block)
       end
 
-      def platform_supports_async?
-        Circuitry.platform_supports_async?
+      def process_via_batch(&block)
+        Processors::Batcher.process(&block)
```

Comment:
I could probably make this more dynamic so it's easier to implement new processors in the future if I wanted to, e.g.:

```ruby
def process_via_async
  Processors.const_get(async.classify).process(&:block)
end
```